File: src/diatune_safe/service.py
from __future__ import annotations

import logging
from datetime import UTC, datetime, timedelta

from diatune_safe.config import AppSettings
from diatune_safe.data_sources import DataSourceError, NightscoutDataSource, SyntheticDataSource
from diatune_safe.domain import AnalysisReport, BlockSettings, PatientProfile, TimeBlock
from diatune_safe.engine import RecommendationEngine
from diatune_safe.repository import SQLiteRepository

logger = logging.getLogger(__name__)


class AnalysisService:
    def __init__(self, settings: AppSettings) -> None:
        self.settings = settings
        self.repository = SQLiteRepository(settings.database_path)
        self.engine = RecommendationEngine(settings)
        self.synthetic_data_source = SyntheticDataSource()
        self.nightscout_data_source = (
            NightscoutDataSource(base_url=settings.nightscout_url, api_secret=settings.nightscout_api_secret)
            if settings.nightscout_url
            else None
        )

    def get_profile(self, patient_id: str) -> PatientProfile:
        profile = self.repository.get_profile(patient_id)
        if profile:
            return profile
        profile = self._default_profile(patient_id)
        self.repository.upsert_profile(profile)
        return profile

    def save_profile(self, patient_id: str, profile: PatientProfile) -> PatientProfile:
        if profile.patient_id != patient_id:
            profile = profile.model_copy(update={"patient_id": patient_id})
        return self.repository.upsert_profile(profile)

    async def run_analysis(
        self,
        *,
        patient_id: str,
        days: int | None = None,
        prefer_real_data: bool = True,
    ) -> AnalysisReport:
        lookback = days if days and days > 0 else self.settings.analysis_lookback_days
        period_end = datetime.now(UTC)
        period_start = period_end - timedelta(days=lookback)
        profile = self.get_profile(patient_id)

        dataset = await self._load_dataset(
            patient_id=patient_id,
            period_start=period_start,
            period_end=period_end,
            prefer_real_data=prefer_real_data,
        )

        report = self.engine.analyze(
            patient_id=patient_id,
            profile=profile,
            dataset=dataset,
            period_start=period_start,
            period_end=period_end,
        )
        saved = self.repository.save_report(report)
        logger.info("Анализ завершен для patient_id=%s run_id=%s", patient_id, saved.run_id)
        return saved

    def get_latest_report(self, patient_id: str) -> AnalysisReport | None:
        return self.repository.get_latest_report(patient_id)

    def list_reports(self, patient_id: str, limit: int = 20) -> list[AnalysisReport]:
        report_ids = self.repository.list_report_ids(patient_id=patient_id, limit=limit)
        reports = [self.repository.get_report(run_id) for run_id in report_ids]
        return [report for report in reports if report is not None]

    def list_pending_recommendations(self, patient_id: str):
        return self.repository.list_pending_recommendations(patient_id)

    def acknowledge_recommendation(self, recommendation_id: int, reviewer: str) -> bool:
        return self.repository.acknowledge_recommendation(recommendation_id=recommendation_id, reviewer=reviewer)

    async def _load_dataset(
        self,
        *,
        patient_id: str,
        period_start: datetime,
        period_end: datetime,
        prefer_real_data: bool,
    ):
        if prefer_real_data and self.nightscout_data_source:
            try:
                dataset = await self.nightscout_data_source.fetch_dataset(
                    patient_id=patient_id,
                    since=period_start,
                    until=period_end,
                )
                if dataset.glucose:
                    return dataset
                logger.warning("Nightscout вернул пустой набор глюкозы; используем синтетические данные.")
            except DataSourceError as exc:
                logger.exception("Ошибка загрузки из Nightscout (%s), используем синтетические данные.", exc)
        return await self.synthetic_data_source.fetch_dataset(
            patient_id=patient_id,
            since=period_start,
            until=period_end,
        )

    def _default_profile(self, patient_id: str) -> PatientProfile:
        return PatientProfile(
            patient_id=patient_id,
            timezone=self.settings.timezone,
            target_low_mgdl=90,
            target_high_mgdl=130,
            blocks=[
                BlockSettings(block=TimeBlock(name="00-03", start_hour=0, end_hour=3), icr=12, isf=55, basal=0.70),
                BlockSettings(block=TimeBlock(name="04-07", start_hour=4, end_hour=7), icr=10, isf=45, basal=0.85),
                BlockSettings(block=TimeBlock(name="08-11", start_hour=8, end_hour=11), icr=9, isf=40, basal=0.80),
                BlockSettings(block=TimeBlock(name="12-15", start_hour=12, end_hour=15), icr=10, isf=42, basal=0.78),
                BlockSettings(block=TimeBlock(name="16-19", start_hour=16, end_hour=19), icr=9, isf=40, basal=0.83),
                BlockSettings(block=TimeBlock(name="20-23", start_hour=20, end_hour=23), icr=11, isf=48, basal=0.72),
            ],
        )

File: internal/service/service.go
package service

import (
	"context"
	"log"
	"time"

	"diatune-safe/internal/config"
	"diatune-safe/internal/datasource"
	"diatune-safe/internal/domain"
	"diatune-safe/internal/engine"
	"diatune-safe/internal/repository"
)

type Service struct {
	settings   config.Settings
	repo       *repository.SQLiteRepository
	engine     engine.Engine
	synthetic  datasource.Source
	nightscout datasource.Source
}

func New(settings config.Settings) (*Service, error) {
	repo, err := repository.NewSQLite(settings.DatabasePath)
	if err != nil {
		return nil, err
	}
	var nightscout datasource.Source
	if settings.NightscoutURL != "" {
		nightscout = datasource.NightscoutSource{
			BaseURL:   settings.NightscoutURL,
			APISecret: settings.NightscoutAPISecret,
			Timeout:   30 * time.Second,
		}
	}
	return &Service{
		settings:   settings,
		repo:       repo,
		engine:     engine.New(settings),
		synthetic:  datasource.SyntheticSource{Seed: 7},
		nightscout: nightscout,
	}, nil
}

func (s *Service) Close() error {
	return s.repo.Close()
}

func (s *Service) GetProfile(patientID string) (domain.PatientProfile, error) {
	p, err := s.repo.GetProfile(patientID)
	if err != nil {
		return domain.PatientProfile{}, err
	}
	if p != nil {
		return *p, nil
	}
	profile := s.defaultProfile(patientID)
	if _, err := s.repo.UpsertProfile(profile); err != nil {
		return domain.PatientProfile{}, err
	}
	return profile, nil
}

func (s *Service) SaveProfile(patientID string, profile domain.PatientProfile) (domain.PatientProfile, error) {
	if profile.PatientID != patientID {
		profile.PatientID = patientID
	}
	return s.repo.UpsertProfile(profile)
}

func (s *Service) RunAnalysis(ctx context.Context, patientID string, days int, preferRealData bool) (domain.AnalysisReport, error) {
	if days <= 0 {
		days = s.settings.AnalysisLookbackDays
	}
	periodEnd := time.Now().UTC()
	periodStart := periodEnd.Add(-time.Duration(days) * 24 * time.Hour)

	profile, err := s.GetProfile(patientID)
	if err != nil {
		return domain.AnalysisReport{}, err
	}
	dataset, _, err := s.loadDatasetWithSource(ctx, patientID, periodStart, periodEnd, preferRealData)
	if err != nil {
		return domain.AnalysisReport{}, err
	}

	report := s.engine.Analyze(patientID, profile, dataset, periodStart, periodEnd)
	saved, err := s.repo.SaveReport(report)
	if err != nil {
		return domain.AnalysisReport{}, err
	}
	log.Printf("анализ завершен patient_id=%s run_id=%v", patientID, saved.RunID)
	return saved, nil
}

func (s *Service) GetLatestReport(patientID string) (*domain.AnalysisReport, error) {
	return s.repo.GetLatestReport(patientID)
}

func (s *Service) ListReports(patientID string, limit int) ([]domain.AnalysisReport, error) {
	ids, err := s.repo.ListReportIDs(patientID, limit)
	if err != nil {
		return nil, err
	}
	out := []domain.AnalysisReport{}
	for _, id := range ids {
		report, err := s.repo.GetReport(id)
		if err != nil {
			return nil, err
		}
		if report != nil {
			out = append(out, *report)
		}
	}
	return out, nil
}

func (s *Service) ListPendingRecommendations(patientID string, limit int) ([]domain.Recommendation, error) {
	if limit <= 0 {
		limit = 100
	}
	return s.repo.ListPendingRecommendations(patientID, limit)
}

func (s *Service) AcknowledgeRecommendation(recommendationID int64, reviewer string) (bool, error) {
	return s.repo.AcknowledgeRecommendation(recommendationID, reviewer)
}

func (s *Service) loadDatasetWithSource(ctx context.Context, patientID string, periodStart, periodEnd time.Time, preferReal bool) (domain.PatientDataset, string, error) {
	if preferReal && s.nightscout != nil {
		dataset, err := s.nightscout.FetchDataset(ctx, patientID, periodStart, periodEnd)
		if err == nil && len(dataset.Glucose) > 0 {
			return dataset, "nightscout", nil
		}
		if err != nil {
			log.Printf("ошибка загрузки из Nightscout (%v), используем синтетические данные", err)
		} else {
			log.Printf("Nightscout вернул пустой набор глюкозы, используем синтетические данные")
		}
	}
	dataset, err := s.synthetic.FetchDataset(ctx, patientID, periodStart, periodEnd)
	if err != nil {
		return domain.PatientDataset{}, "", err
	}
	return dataset, "synthetic", nil
}

func (s *Service) loadDataset(ctx context.Context, patientID string, periodStart, periodEnd time.Time, preferReal bool) (domain.PatientDataset, error) {
	dataset, _, err := s.loadDatasetWithSource(ctx, patientID, periodStart, periodEnd, preferReal)
	return dataset, err
}

func (s *Service) defaultProfile(patientID string) domain.PatientProfile {
	return domain.PatientProfile{
		PatientID:      patientID,
		Timezone:       s.settings.Timezone,
		TargetLowMgdl:  90,
		TargetHighMgdl: 130,
		Blocks: []domain.BlockSettings{
			{Block: domain.TimeBlock{Name: "00-03", StartHour: 0, EndHour: 3}, ICR: 12, ISF: 55, Basal: 0.70},
			{Block: domain.TimeBlock{Name: "04-07", StartHour: 4, EndHour: 7}, ICR: 10, ISF: 45, Basal: 0.85},
			{Block: domain.TimeBlock{Name: "08-11", StartHour: 8, EndHour: 11}, ICR: 9, ISF: 40, Basal: 0.80},
			{Block: domain.TimeBlock{Name: "12-15", StartHour: 12, EndHour: 15}, ICR: 10, ISF: 42, Basal: 0.78},
			{Block: domain.TimeBlock{Name: "16-19", StartHour: 16, EndHour: 19}, ICR: 9, ISF: 40, Basal: 0.83},
			{Block: domain.TimeBlock{Name: "20-23", StartHour: 20, EndHour: 23}, ICR: 11, ISF: 48, Basal: 0.72},
		},
	}
}
